### ISSUE TITLE
Intermittent error 7 when creating a google_service_networking_connection resource

### ISSUE THREAD

**Issue by lemoo5sg** [hashicorp/terraform-provider-google#14720]
### Description

When creating a google_service_networking_connection resource, we intermittently face the “ Error code 7, message: Required 'compute.globalAddresses.list' permission” error.

** google_service_networking_connection resource creation**:
```tf
resource "google_compute_global_address" "psa_ranges" {
  for_each      = var.psa_config.ranges
  project       = var.project_id
  name          = each.key
  purpose       = "VPC_PEERING"
  address_type  = "INTERNAL"
  address       = split("/", each.value)[0]
  prefix_length = split("/", each.value)[1]
  network       = local.network.id
}

resource "google_service_networking_connection" "psa_connection" {
  network  = local.network.id
  service  = "servicenetworking.googleapis.com"
  reserved_peering_ranges = [
    for k, v in google_compute_global_address.psa_ranges : v.name
  ]
}
```

We have found the posts [1] and [2] recommending in this case to either disable & re-enable the api and/or to grant the servicenetworking.serviceAgent role to the service networking api service account.

example:
```tf
resource "google_project_iam_member" "servicenetworking_role" {
  project = var.project_id
  role    = "role/servicenetworking.serviceAgent"
  member  = "serviceAccount:service-xxx@service-networking.iam.gserviceaccount.com"
}
```

However, after implementing this workaround, we could create other projects, however we still have had some cases where the respective service account role grant is failing because the service account does not exist yet for these projects though the Service Networking API is enabled.

Error:
```
Error: Request `Create IAM Members roles/servicenetworking.serviceAgent serviceAccount:service-xxx@service-networking.iam.gserviceaccount.com for project "xxx"` returned error: Error applying IAM policy for project "xx": 
Error setting IAM policy for project "xxxx": googleapi: Error 400: Service account service-yyy@service-networking.iam.gserviceaccount.com does not exist., badRequest

  with google_project_iam_member.servicenetworking_role,
  on main.tf line xxx, in resource "google_project_iam_member" "servicenetworking_role":
```

According to the GCP documentation [3],  the service networking account provisioning will happen just-in-time. This means that the service account will only be created when there is an api call and once the service account gets provisioned, it is created for the project.

Our question is: would you know with Terraform how to force touch this api call to make sure the service-networking api service account will always be created before granting the permission ?
Is there any other resource we could create for the same ?

### New or Affected Resource(s)

* google_service_networking_connection

### References

Reference documentation:

* [1]  https://github.com/hashicorp/terraform-provider-google/issues/11195
* [2] https://thedataguy.in/cloudsql-shared-vpc-private-ip-and-servicenetworking.serviceagent-role/
* [3] https://cloud.google.com/service-infrastructure/docs/service-networking/getting-started

**Comment by edwardmedia** [comment]
@lemoo5sg can you hardcode the config first before you add variables and dynamic code? This will make explanation easier.

**Comment by lemoo5sg** [comment]
> @lemoo5sg can you hardcode the config first before you add variables and dynamic code? This will make explanation easier.

Hi @edwardmedia description is updated, tried to hardcode , not sure about the reserved_peering_ranges expected value here, hope this helps

**Comment by edwardmedia** [comment]
@lemoo5sg not sure if you have had chance to review [this
](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/provider_reference)

To provision `google_service_networking_connection`, it does have many dependencies. For that, you may review this [doc](https://cloud.google.com/vpc/docs/configure-private-services-access) and its above to find out what they are. Remember, to build each resource, you may need to enable its API and give related IAMs first. 

For details, you may need to review individual document.  

Maybe you can start with methods like gcloud by following the [doc](https://cloud.google.com/vpc/docs/configure-private-services-access) to see if that makes sense?

**Comment by lemoo5sg** [comment]
@edwardmedia Indeed, to provision this resource a lot of other dependancies are needed. The provided code is only a very small part of our landing zone full structure and actually these resources are the one created through the [Google  Cloud Foundation Fabric net-vpc](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/modules/net-vpc/psa.tf)) module.

We have mentioned that the issue is intermittent, for many of the projects we create, the google_service_networking_connection is successfully created with the expected dependant resources, however often it fails as described.

The first workaround we have used is described [here](https://github.com/hashicorp/terraform-provider-google/issues/11195) and [here](https://thedataguy.in/cloudsql-shared-vpc-private-ip-and-servicenetworking.serviceagent-role/). This workaround has helped for a lot of projects, however still with have the second described issue for some remaining projects.

**Comment by edwardmedia** [comment]
@lemoo5sg Cool. What are the remaining projects? In this case, it is likely some behavior on the API side. There are not much the Terraform provider can do. My suggestion would be try to set it up with gcloud to see if you can make it working. , which could help you identify where the problem is. 

```
however still with have the second described issue for some remaining projects.
```

Does this make sense?

**Comment by lemoo5sg** [comment]
For reference, we are testing the solution provided [here](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/issues/1399#event-9347633834) by the foundation fabric team. The google_project_service_identity seems to be the missing piece for us.

**Comment by edwardmedia** [comment]
@lemoo5sg cool. any remaining issues?

**Comment by lemoo5sg** [comment]
We have implemented the fix and we will create new projects next week. To be continued...
Closing this case anyway.
